README.md
# Revealing Snow

Revealing Snow is a web-based software platform for interactively visualizing the impact
of varying screening criteria on patient counts. The tool also allows candidates
to be selected based on YMCA/facility proximity for export to the recruitment
and mail tracking system.


## License

The most appropriate open source license for this project has not yet been determined.

Until a license decision has been made, the authors retain exclusive copyright to this software.


CHANGELOG.md
# Revealing Snow Change Log

## [2020.2.0] - 2020-11-04

### Added

* It's now possible to filter subjects by demographic fields (sex, race, and ethnicity).
  Multiple values can be selected, and only the selected values will be included in
  the "eligible" population.

## [2020.1.0] - 2020-03-09

### Added

* Added the Bryan and Spears Greensboro YMCA sites.
* Added last-updated dates for the EMR Data Pull and the "Previously Exported" column.

### Changed

* Reduced the set of default filter criteria to a minimal starting set of conditions.
* The charts in the Patient Totals section now use the same category colors and ordering
  as the site-specific charts.
* When one or more YMCA sites is selected, the "Patient Totals" now reflect the union
  of the patients in all selected sites.


## [2019.5.2] - 2020-02-04

### Added

* Added a filter indicating whether a patient was in a previous export batch.

### Changed

* Updated the data version in the model.
* Corrected the vitals item description to remove BMI.

### Fixed

* Renamed the current_age column to match the screening table.


## [2019.5.1] - 2020-01-09

### Added

* Added the Thomasville site.

### Changed

* Updated descriptions in the data model to reflect POINTER Protocol v2.0.

## [2019.5.0] - 2019-12-31

### Changed

* Relabeled "Unfiltered" and "Filtered" to "Overall" and "Eligible."
* Added support for data files without the redundant "_date" columns, such
  as those produced by v4 of the phenotyping algorithm.

### Fixed

* Icons in the filter legend popover appear again.

## [2019.4.0] - 2019-09-26

### Added

* Added the High Point Medical Center fitness center.

## [2019.3.0] - 2019-08-28

### Added

* Added the Lexington J. Smith Young YMCA.

## [2019.2.0] - 2019-06-06

### Changed

* Updated the model to use the v3 screening criteria.
* Default dates are now relative to the 2019-04-01 reference date.
* Range filters (e.g. age) now support non-integer values.
* Exclusion filters now specifically exclude patients with condition = 1
  rather than requiring condition = 0. This ensures empty/unknown condition
  values are not excluded.

## [2019.1.0] - 2019-02-15

### Added

* The YMCA site distance filters now support specifying a minimum distance.

## [2018.6.2] - 2019-01-03

### Added

* The download and export payload now includes the data version string.

### Changed

* Changed distance cutoffs to use <= instead of <.
* Updated the data model to allow newer versions of the screening data to be used.

## [2018.6] - 2018-12-28

### Added

* The export dialog now has a field for user name.
* The app can now be built and distributed as a standalone Windows app.

### Changed

* Exporting to remote tracking is now a toggleable feature. It defaults to being turned off.

### Fixed

* Removed the artificial max of 1000 from the limit field in the filter bar.


## [2018.5] - 2018-12-14

### Added

* Metadata from previous exports can now be re-imported.
* Added age to the available filter criteria.

### Fixed

* Sequential load and unload of chart data results in incorrect chart appearance.


## [2018.4] - 2018-10-17

### Added

* Value tooltips now include percentages.
* Data exports now support labels and descriptions.
* Patient subsetting also supports ordering by YMCA distance.

### Changed

* Charts have more complete annotations.


## [2018.3] - 2018-10-03

### Added

* Data can now be exported to the remote mail tracking system in addition to downloading locally.
* The set of patients downloaded/exported can now be limited and sorted by last visit date.

### Fixed

* Inconsistent color scheme when groups are different between charts.
* Demographic charts don't update when categories are empty.


## [2018.2] - 2018-09-07

### Added

* Metadata and YMCA site cutoffs are included in exported data.
* Added a legend to explain inclusion/exclusion/ignore icons.
* Demographic data are now included among the YMCA site charts.

### Changed

* Modifications to filter or site criteria are now applied automatically.
* Filter criteria are now three-way checkboxes.
* Cutoff dates can now be entered by keyboard.

### Fixed

* Distances are now rounded correctly.
* Filter controls are no longer unclickable on narrow screens.


## [2018.1] - 2018-05-18

* Initial Release


api/snow/app.py
import logging.config
import os

import yaml
from flask import Flask, request

from snow import constants as C


# Adapted From: https://fangpenlin.com/posts/2012/08/26/good-logging-practice-in-python/
def _setup_logging(configfile, default_level=logging.INFO):
    if os.path.exists(configfile):
        with open(configfile, 'rt') as f:
            config = yaml.safe_load(f.read())
            logging.config.dictConfig(config)
    else:
        logging.basicConfig(level=default_level)


def log_request_info():
    import logging
    logger = logging.getLogger(__name__)

    logger.debug(
        "%s - %s - %s",
        request.remote_addr,
        request.method,
        request.url
    )


def create_app(configuration=None):
    from snow import config
    from snow import query
    from snow import model
    from snow import export
    from snow import features

    from snow.tracking import tracking
    from snow.ptscreen import pscr

    app = Flask(__name__)
    app.config.from_object(configuration or config.Configuration)

    _setup_logging(app.config[C.LOGGING_CONFIG_FILE])
    app.before_request(log_request_info)

    pscr.init_app(app)
    model.cdm.init_app(app)
    tracking.init_app(app)

    app.add_url_rule('/', view_func=lambda: app.send_static_file('index.html'))
    app.add_url_rule('/stats', 'stats', query.patient_stats)
    app.add_url_rule('/download', 'download', export.download_patients)
    app.add_url_rule('/export', 'export', export.export_patients, methods=['POST'])
    app.add_url_rule('/ymca_stats', 'ymca_stats', query.ymca_stats)
    app.add_url_rule('/model', 'model', model.get_criteria_data_model)
    app.add_url_rule('/features', 'features', features.get_feature_flags)

    return app


def main():
    create_app().run('0.0.0.0', 5000)


def runtornado():
    from snow import wsgi, util
    from tornado import wsgi as twsgi
    from tornado import httpserver
    from tornado import ioloop
    from tornado import netutil

    container = twsgi.WSGIContainer(wsgi.app)
    sockets = netutil.bind_sockets(0, '')

    http_server = httpserver.HTTPServer(container)
    http_server.add_sockets(sockets)

    util.open_browser(sockets)
    ioloop.IOLoop.current().start()
